### PR TITLE
Accomodate Chrome m75 WebXR API changes. Function renaming.

### PR DESCRIPTION
**Pull request by dmarcos**
This covers latest changes landed in Chrome Canary to keep WebXR / WebVR working in all browsers.

There are more API changes coming. e.g: [XRDevice API will be gone](https://github.com/immersive-web/webxr/commit/6acdaa33e1d421f4ff445d41ce7270f787006f46)

Tested on: 

- Pixel 3 (Daydream), Chrome m71 and m73 with WebXR flags enabled
- Chrome Desktop m71 and m73 on Oculus with WebXR flags enabled.
- Firefox 64 Desktop (WebVR API) on Oculus.
- Supermedium (WebVR API) on Oculus

@jsantell Thanks, I used you [PR](https://github.com/GoogleWebComponents/model-viewer/blob/master/src/three-components/ARRenderer.js#L110) as reference.

### CODE REVIEW

File: examples/webvr_cubes.html [hunk 1]
```diff
@@ -60,8 +60,11 @@
 				scene = new THREE.Scene();
 				scene.background = new THREE.Color( 0x505050 );
 
+				var dolly = new THREE.Group();
+				dolly.position.set(0, 1.6, 0);
```
Comment (jsantell) [inline examples/webvr_cubes.html]: Shouldn't this be set with reference frame or WebVR's standing matrix?

File: src/renderers/webvr/WebXRManager.js [hunk 2]
```diff
@@ -215,18 +241,18 @@ function WebXRManager( renderer ) {
...
 	function onAnimationFrame( time, frame ) {
 
-		pose = frame.getDevicePose( frameOfReference );
+		pose = frame.getDevicePose !== undefined ? frame.getDevicePose( frameOfReference ) : frame.getViewerPose( frameOfReference );
 
 		if ( pose !== null ) {
 
 			var layer = session.baseLayer;
-			var views = frame.views;
+			var views = frame.views || pose.views;
```
Comment (jsantell) [inline src/renderers/webvr/WebXRManager.js]: TIL XRViewerPose has added `views` and XRView has added `viewMatrix` earlier this week

**Review by jsantell — COMMENTED** [review]
LGTM! Although, with Chrome m73 starting to land this 'inbetween' state, I'm not sure how valuable it will be to keep up with the changes until there's more stability. Some comments here (https://github.com/GoogleWebComponents/model-viewer/pull/292) about plans on handling the transition, although that's a bit more downstream than three.js

File: examples/webvr_cubes.html (hunk 1 above)
Comment (dmarcos) [reply examples/webvr_cubes.html]: True, it should be handled on the WebVR / WebXR side but couldn't find a way to get consistent behavior across all API flavors. Specially with the m73 changes. I can remove to keep this simpler.

**Comment by dmarcos** [comment]
Document provided by Google with info about WebXR changes:

https://docs.google.com/document/d/1R8Bz0_vK9YmY5-ZhsbLtZ9-DDJpTOgNeibFD4SdUnbw/edit?usp=sharing

File: src/renderers/webvr/WebXRManager.js [hunk 3]
```diff
@@ -77,7 +77,19 @@ function WebXRManager( renderer ) {
...
 		if ( value !== undefined ) device = value;
-		if ( value instanceof XRDevice ) gl.setCompatibleXRDevice( value );
+		if ( value instanceof XRDevice ) {
+
+			if ( gl.setCompatibleXRDevice !== undefined ) {
+
+				gl.setCompatibleXRDevice( value );
+
+			} else {
+
```
Comment (mrdoob) [inline src/renderers/webvr/WebXRManager.js]: Do you mind adding this comment here (and in all the other `else`s)?

```
// TODO: Remove when Chrome 73 reaches stable
```

Comment (dmarcos) [reply src/renderers/webvr/WebXRManager.js]: @mrdoob Yep, I can accomodate / test latest Canary changes. Do you want to merge? @jsantell recommended to maybe wait until the dust settles. It's true there's a fair amount of churn at the moment. I let you decide.

**Comment by mrdoob** [comment]
Hmm, yeah. Maybe better to wait until things get stable.
Better to focus on Oculus Go/Quest until then.

File: src/renderers/webvr/WebXRManager.js [hunk 4]
```diff
@@ -121,16 +144,19 @@ function WebXRManager( renderer ) {
 			session.addEventListener( 'end', onSessionEnd );
 
 			session.baseLayer = new XRWebGLLayer( session, gl, { framebufferScaleFactor: framebufferScaleFactor } );
```
Comment (mvilledieu) [inline src/renderers/webvr/WebXRManager.js]: Add new support for updateRenderState() to get the session's `baseLayer`.

	if ( session.updateRenderState !== undefined ) {

		session.updateRenderState( {  baseLayer:  new XRWebGLLayer( session, gl, {  framebufferScaleFactor: framebufferScaleFactor } ) } );

	} else {

		session.baseLayer =  new XRWebGLLayer( session, gl, {  framebufferScaleFactor: framebufferScaleFactor } );

	}

File: src/renderers/webvr/WebXRManager.js [hunk 5]
```diff
@@ -215,18 +241,18 @@ function WebXRManager( renderer ) {
 
 	function onAnimationFrame( time, frame ) {
 
-		pose = frame.getDevicePose( frameOfReference );
+		pose = frame.getDevicePose !== undefined ? frame.getDevicePose( frameOfReference ) : frame.getViewerPose( frameOfReference );
 
 		if ( pose !== null ) {
 
 			var layer = session.baseLayer;
```
Comment (mvilledieu) [inline src/renderers/webvr/WebXRManager.js]: Same here with
`var layer =  'renderState' in session ? session.renderState.baseLayer : session.baseLayer;`

File: src/renderers/webvr/WebXRManager.js [hunk 6]
```diff
@@ -97,6 +109,17 @@ function WebXRManager( renderer ) {
 
 	}
 
+	function onRequestFrameOfReference( value ) {
+
+		frameOfReference = value;
+
+		renderer.setFramebuffer( session.baseLayer.framebuffer );
```
Comment (mvilledieu) [inline src/renderers/webvr/WebXRManager.js]: And here
`renderer.setFramebuffer( 'renderState' in session ? session.renderState.baseLayer.framebuffer : session.baseLayer.frameBuffer );`

**Review by mvilledieu — CHANGES_REQUESTED** [review]
Add support for `session.updateRenderState()` to get the session's `baseLayer`.

File: src/renderers/webvr/WebXRManager.js (hunk 4 above)
Comment (Azier777) [reply src/renderers/webvr/WebXRManager.js]: https://github.com/mrdoob/three.js/pull/15604#discussion_r257491150

File: src/renderers/webvr/WebXRManager.js (hunk 3 above)
Comment (Azier777) [reply src/renderers/webvr/WebXRManager.js]: https://github.com/mrdoob/three.js/pull/15604#discussion_r257491150

**Review by Azier777 — APPROVED** [review]

**Comment by dmarcos** [comment]
@mvilledieu Thanks for keeping an eye on this. I incorporated changes and verified webvr examples present on Oculus Rift with Firefox 65. Chrome m72 and m74 (Canary)

**Comment by dmarcos** [comment]
Changes made their way to m73 stable and WebXR is broken at the moment including WebXR samples. Implementation doesn't seem to match spec and couldn't find a working code sample to fix this PR.

File: src/renderers/webvr/WebXRManager.js (hunk 2 above)
Comment (klausw) [reply src/renderers/webvr/WebXRManager.js]: FYI, looks like `viewMatrix` is getting removed again, instead the spec is adding an `.inverse()` method (or `.inverse` attribute?) to the pose XRRigidTransform: 

https://github.com/immersive-web/webxr/pull/560/files

**Comment by dmarcos** [comment]
Added commit to bring compatibility with m75 (Canary). Thanks to @klausw help.

File: src/renderers/WebGLRenderer.js [hunk 7]
```diff
@@ -2418,9 +2419,10 @@ function WebGLRenderer( parameters ) {
 	}
 
 	//
-
 	this.setFramebuffer = function ( value ) {
 
+		if (_framebuffer !== value ) _gl.bindFramebuffer( _gl.FRAMEBUFFER, value );
```
Comment (dmarcos) [inline src/renderers/WebGLRenderer.js]: Change submitted as a separate PR: https://github.com/mrdoob/three.js/pull/16118

**Comment by cabanier** [comment]
Since the API is still changing and different versions of chrome might be out in the field, can you add a strict check for the version of chrome that you know works?
If not running that version, revert to WebVR.

**Comment by dmarcos** [comment]
Likely best is to always use Canary as point of reference until the implementation stabilizes.

**Comment by mrdoob** [comment]
Thanks!
